src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -67,23 +199,57 @@ protected override void EndProcessing()
         {
             OrderByProperty orderByProperty = new OrderByProperty(
                 this, InputObjects, Property, !Descending, ConvertedCulture, CaseSensitive);
-            if (orderByProperty.Comparer == null || orderByProperty.OrderMatrix == null || orderByProperty.OrderMatrix.Count == 0)
+
+            var dataToProcess = orderByProperty.OrderMatrix;
+            var comparer = orderByProperty.Comparer;
+            if (comparer == null || dataToProcess == null || dataToProcess.Count == 0)
             {
                 return;
             }
-            orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
 
-            if (_unique)
+            // Track the range of sorted data in the data that we are processing
+            int outputStartIndex = 0;
+            int outputEndIndex = dataToProcess.Count - 1;
+
+            // If -Unique was used, if -Top & -Bottom were not used, or if -Top or -Bottom would return all objects, sort
+            // all objects, remove duplicates if necessary, and identify the list of items to return
+            if (_unique || (Top == 0 && Bottom == 0) || Top >= dataToProcess.Count || Bottom >= dataToProcess.Count)
```

Comment:
Same comment about -Unique and heapify applies here. I'm testing something with it though to see what impact it might have.